android/libraries/rib-badoo/src/main/java/com/badoo/common/rib/routing/backstack/BackStackManager.kt

```diff
@@ -0,0 +1,186 @@
+package com.badoo.common.rib.routing.backstack
+
+import android.os.Parcelable
+import com.badoo.common.rib.routing.RibConnector
+import com.badoo.common.rib.routing.action.RoutingAction
+import com.badoo.common.rib.routing.backstack.BackStackManager.Action
+import com.badoo.common.rib.routing.backstack.BackStackManager.Action.Execute
+import com.badoo.common.rib.routing.backstack.BackStackManager.Action.ReattachRibsOfLastEntry
+import com.badoo.common.rib.routing.backstack.BackStackManager.Effect
+import com.badoo.common.rib.routing.backstack.BackStackManager.State
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.NewRoot
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.Pop
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.Push
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.Replace
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.ShrinkToBundles
+import com.badoo.common.rib.routing.backstack.BackStackManager.Wish.TearDown
+import com.badoo.common.rib.routing.backstack.BackStackRibConnector.DetachStrategy.DESTROY
+import com.badoo.common.rib.routing.backstack.BackStackRibConnector.DetachStrategy.DETACH_VIEW
+import com.badoo.mvicore.element.Actor
+import com.badoo.mvicore.element.Bootstrapper
+import com.badoo.mvicore.element.Reducer
+import com.badoo.mvicore.element.TimeCapsule
+import com.badoo.mvicore.feature.BaseFeature
+import io.reactivex.Observable
+import io.reactivex.Observable.empty
+import io.reactivex.Observable.just
+import kotlinx.android.parcel.Parcelize
+
+internal class BackStackManager<C : Parcelable>(
+    resolver: (C) -> RoutingAction<*>,
+    ribConnector: RibConnector,
+    initialConfiguration: C,
+    timeCapsule: TimeCapsule<State<C>>,
+    tag: String
+): BaseFeature<Wish<C>, Action<C>, Effect<C>, State<C>, Nothing>(
+    initialState = timeCapsule[tag] ?: State(),
+    wishToAction = { Execute(it) },
+    bootstrapper = BooststrapperImpl(timeCapsule[tag] ?: State(), initialConfiguration),
+    actor = ActorImpl<C>(
+        connector = BackStackRibConnector(
+            resolver,
+            ribConnector
+        )
+    ),
+    reducer = ReducerImpl<C>()
+) {
+    init {
+        timeCapsule.register(tag) { state }
+    }
+
+    @Parcelize
+    data class State<C : Parcelable>(
+        val backStack: List<BackStackElement<C>> = emptyList()
+    ) : Parcelable {
+        val current: C
+            get() = backStack.last().configuration
+
+        val canPop: Boolean
+            get() = backStack.size > 1
+    }
+
+    class BooststrapperImpl<C : Parcelable>(
+        private val state: State<C>,
+        private val initialConfiguration: C
+    ) : Bootstrapper<Action<C>> {
+        override fun invoke(): Observable<Action<C>> = when {
```

Comment:
You can just emit action like "Init" and move this logic into Actor. Then you don't need to tricky-pass state and configuration.

Reply:
Good idea, thanks. Wait, doesn't that mean that the NewRoot part of the when expression in the Actor would need to be copied under the Init branch? I think that would result in duplication.